sklearn/model_selection/_classification_threshold.py

```diff
@@ -0,0 +1,1189 @@
+                )
+
+            self.estimator_.fit(X_train, y_train, **fit_params_train)
+
+        cv_thresholds, cv_scores = zip(
+            *Parallel(n_jobs=self.n_jobs)(
+                delayed(_fit_and_score_over_thresholds)(
+                    clone(classifier) if cv != "prefit" else classifier,
+                    X,
+                    y,
+                    fit_params=routed_params.estimator.fit,
+                    train_idx=train_idx,
+                    val_idx=val_idx,
+                    curve_scorer=self._curve_scorer,
+                    score_params=routed_params.scorer.score,
+                )
+                for train_idx, val_idx in splits
+            )
+        )
+
+        if any(np.isclose(th[0], th[-1]) for th in cv_thresholds):
+            raise ValueError(
+                "The provided estimator makes constant predictions. Therefore, it is "
+                "impossible to optimize the decision threshold."
+            )
+
+        # find the global min and max thresholds across all folds
+        min_threshold = min(
+            split_thresholds.min() for split_thresholds in cv_thresholds
+        )
+        max_threshold = max(
+            split_thresholds.max() for split_thresholds in cv_thresholds
+        )
+        if isinstance(self.thresholds, Integral):
+            decisiothresholds = np.linspace(
+                min_threshold, max_threshold, num=self.thresholds
+            )
+        else:
+            decisiothresholds = np.asarray(self.thresholds)
+
+        if not constrained_metric:  # find best score that is the highest value
+            objective_scores = _mean_interpolated_score(
+                decisiothresholds, cv_thresholds, cv_scores
+            )
+            best_idx = objective_scores.argmax()
+            self.best_score_ = objective_scores[best_idx]
+            self.best_threshold_ = decisiothresholds[best_idx]
+            self.constrained_score_ = None
+            if self.store_cv_results:
+                self.cv_results_ = {
+                    "thresholds": decisiothresholds,
+                    "scores": objective_scores,
+                }
+        else:
+            if "tpr" in self.objective_metric:  # tpr/tnr
+                mean_tnr, mean_tpr = [
+                    _mean_interpolated_score(decisiothresholds, cv_thresholds, sc)
+                    for sc in zip(*cv_scores)
+                ]
+            else:  # precision/recall
+                mean_precision, mean_recall = [
+                    _mean_interpolated_score(decisiothresholds, cv_thresholds, sc)
+                    for sc in zip(*cv_scores)
+                ]
+
+            def _get_best_idx(constrained_score, maximized_score):
+                """Find the index of the best score constrained by another score."""
+                mask = constrained_score >= self.constraint_value
+                mask_idx = maximized_score[mask].argmax()
```

Comment:
Around here is the part which can be simplified if we deal with callables as a constraint and have different pre-defined ones instead of strings.